flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskLocalStateStoreImpl.java

```diff
@@ -141,6 +158,29 @@ public void storeLocalState(
                     subtaskIndex);
         }
 
+        for (Map.Entry<OperatorID, OperatorSubtaskState> subtaskStateEntry :
+                localState.getSubtaskStateMappings()) {
+            for (KeyedStateHandle keyedStateHandle :
+                    subtaskStateEntry.getValue().getManagedKeyedState()) {
+                if (keyedStateHandle instanceof ChangelogStateBackendHandle) {
+                    ChangelogStateBackendHandle changelogStateBackendHandle =
+                            (ChangelogStateBackendHandle) keyedStateHandle;
+                    long materializationID = changelogStateBackendHandle.getMaterializationID();
+                    if (currentMaterializationID == null
```

Comment:
This assumes that the snapshot contains at most one materialization ID.
Can (and should) we rely on it? 
I'm not sure whether it holds for down-scaling case.

Reply:
1. Local recovery and rescaling can not happen at the same time, we can temporarily assume that the snapshot contains at most one materialization ID now.
2. Even if local recovery and rescaling can happen at the same time later,  `storeLocalState()` is called after   restoring is complete, and there is also at most one materialization ID per `ChangelogStateBackendHandle`, because we choose max materializationID to handle rescaling.
https://github.com/apache/flink/blob/master/flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogKeyedStateBackend.java#L616-L617

Reply:
1. Yes, but this method is called after the actual recovery - when reporting a new checkpoint
2. Yes, but relying on it seems fragile

How about adding an explicit check like this:
```
                    if (mapToMaterializationId.containsKey(checkpointId)) {
                        checkState(materializationID == mapToMaterializationId.get(checkpointId));
                    } else {
                        mapToMaterializationId.put(checkpointId, materializationID);
                    }
```

and removing 
```
if (mapToMaterializationId.getOrDefault(checkpointId, Long.MAX_VALUE)
        < materializationID) {
    LOG.info(
```
?